email: Open a single SMTP connection to send email batches.

pletinckxc opened the pull request:
Previously the outgoing emails were sent over several SMTP
connections through the EmailSendingWorker. In order to improve
the efficiency of this process the EmailSendingWorker worker is
now defined based on the LoopQueueProcessingWorker which allows to
handle batches of events from the queue workers. This permits to
open a single SMTP connection to send the whole batch of emails.

This was implemented directly in the consume_batch function
inherited from the LoopQueueProcessingWorker's abstract method.
This concrete implementation first creates and opens a new SMTP
connection from the default Django SMTP server then sends each
email of the batch through this connection and finally closes the
connection.

A first retry mechanism is used to redo the opening of the
connection until it works or until three attempts failed. A second
retry mechanism tackles errors that could happen when sending a
particular email from the batch and for which the process is
retriable (e.g. an SMTPServerDisconnected error). If the error
doesn't fall in this category then the particular email is just
dropped (e.g. an SMTPRecipientsRejected).

**Help/Feedback required**

This second part about the retry mechanisms is trickier than what
I first thought and I must say I am a bit confused about how I
could adapt the "retry_send_email_failures" decorator to be able
to use it in the new consume_batch function. I'm then open to any
suggestions to improve the current workaround that I tried to
implement but that still makes the backend tests fail.

**Testing plan:** <!-- How have you tested? -->

Automatic tests.

Fixes: #17672.

alexmv commented on zerver/worker/queue_processors.py:
```diff
@@ -607,17 +608,44 @@ def maybe_send_batched_emails(self) -> None:
+                handle_send_email_format_changes(copied_event)
+                send_email(self, copied_event)
+            try:
+                connection.close()
```
We explicitly _don't_ want to close the connection after every email we send.

pletinckxc replied:
Indeed, my bad there. Instead I overwrote the stop() function of EmailSendingWorker to close the connection when we stop the queue.

alexmv commented on zerver/worker/queue_processors.py:
```diff
@@ -607,17 +608,44 @@ def maybe_send_batched_emails(self) -> None:
+            send_email_from_dict(event)
+
+        try:
+            connection = initialize_connection()
```
We ideally want this connection to be a property of the `EmailSendingWorker` so we can possibly even reuse it across batches.

With the definition of `initialize_connection` earlier, we want something like:

```py3
self.connection = initialize_connection(self.connection)
```

pletinckxc replied:
I overwrote the __init()__ method of EmailSendingWorker to include this change and I created a new method for the send_email() function that you mentioned in the previous review message.

alexmv commented on zerver/worker/queue_processors.py:
```diff
@@ -607,17 +608,44 @@ def maybe_send_batched_emails(self) -> None:
-        send_email_from_dict(copied_event)
+class EmailSendingWorker(LoopQueueProcessingWorker):
+    def consume_batch(self, events: List[Dict[str, Any]]) -> None:
+        @retry_send_email_failures
+        def send_email(worker: EmailSendingWorker, event: Dict[str, Any]) -> None:
+            send_email_from_dict(event)
```
I think this will be less weird if you make it a method on the `EmailSendingWorker`, and not a nested function.

alexmv replied:
But see below for the problem about this method.

pletinckxc replied:
Answer below.

alexmv commented on zerver/worker/queue_processors.py:
```diff
@@ -607,17 +608,44 @@ def maybe_send_batched_emails(self) -> None:
+
+        try:
+            connection = initialize_connection()
+        except OSError as e:
+            self._handle_consume_exception(events, e)
```
This is pretty much equivalent to just letting the exception bubble up -- which would be fine, and probably easier to understand.

pletinckxc replied:
Yes, modified.

alexmv commented on zerver/worker/queue_processors.py:
```diff
@@ -607,17 +608,44 @@ def maybe_send_batched_emails(self) -> None:
+# class EmailSendingWorker(QueueProcessingWorker):
+#     @retry_send_email_failures
+#     def consume(self, event: Dict[str, Any]) -> None:
+#         # Copy the event, so that we don't pass the `failed_tries'
+#         # data to send_email_from_dict (which neither takes that
+#         # argument nor needs that data).
```
This comment, about _why_ we do the deepcopy, would be good to preserve in some form.

pletinckxc replied:
The comment has been reintroduced.

alexmv reviewed (commented):
This is a hard, complicated problem!  Thank you for digging into it.

I've left a bunch of comments inline -- let me know if they clarify things, or any further questions that they bring to mind.

pletinckxc commented on zerver/lib/send_email.py:
```diff
@@ -24,6 +25,8 @@
 from zerver.lib.logging_util import log_to_file
 from zerver.models import EMAIL_TYPES, Realm, ScheduledEmail, UserProfile, get_user_profile_by_id
 
+MAX_CONNECTION_TRIALS = 3
```
Sure, modified.

pletinckxc commented on zerver/lib/send_email.py:
```diff
@@ -210,6 +221,7 @@ def send_email(
         language=language,
         context=context,
         realm=realm,
+        connection=connection,
```
See above.

pletinckxc commented on zerver/lib/send_email.py:
```diff
@@ -223,6 +235,17 @@ def send_email_from_dict(email_dict: Mapping[str, Any]) -> None:
     send_email(**dict(email_dict))
 
 
+def initialize_connection() -> smtplib.SMTP:
+    connection = get_connection()
+    for iter in range(MAX_CONNECTION_TRIALS):
+        try:
+            connection.open()
+            return connection
```
Backoff is a great library I didn't know it, thank you for sharing. I inserted the dependency in the requirements files and I modified the function to include the on_exception decorator with the max_tries argument but I'm not sure about what I need to add to include the timeout. Is it already done ?

Checking the instance of the connection to ensure that the remote server didn't close the connection is also a great idea ! I modified the function to implement these changes as well.

For the OSError I digged a a bit into the django source code and it seems that it is the unique error that can be raised by [open()](https://github.com/django/django/blob/a815a6a3159d5f708866ca968e8994f201ea3e43/django/core/mail/backends/smtp.py#L41). An ImportError can also happen from the [get_connection()](https://github.com/django/django/blob/a815a6a3159d5f708866ca968e8994f201ea3e43/django/core/mail/__init__.py#L26) call through [import_string()](https://github.com/django/django/blob/a815a6a3159d5f708866ca968e8994f201ea3e43/django/utils/module_loading.py#L7) but I'm not sure we want to catch that as it shouldn't happen right ?

alexmv replied:
> I inserted the dependency in the requirements files and I modified the function to include the on_exception decorator with the max_tries argument but I'm not sure about what I need to add to include the timeout. Is it already done ?

Yup; see https://github.com/zulip/zulip/blob/1a8ad796f89d9c04f3fa94e6a8b45a79e42f7067/zproject/computed_settings.py#L1132 and https://docs.djangoproject.com/en/3.2/ref/settings/#std:setting-EMAIL_TIMEOUT.

> For the OSError I digged a a bit into the django source code and it seems that it is the unique error that can be raised by open(). An ImportError can also happen from the get_connection() call through import_string() but I'm not sure we want to catch that as it shouldn't happen right ?

Yeah, retrying `ImportError` is not going to result in a better outcome later. :). `OSError` covers [`socket.gaierror`](https://docs.python.org/3/library/socket.html#socket.gaierror) and [`socket.timeout`](https://docs.python.org/3/library/socket.html#socket.timeout) which are the two that `retry_send_email_failures` specifically retry -- so I think that should suffice.

pletinckxc commented:
> This is a hard, complicated problem! Thank you for digging into it.
> 
> I've left a bunch of comments inline -- let me know if they clarify things, or any further questions that they bring to mind.

Yes indeed at least harder than I thought in the first place but your comments helped me to learn a lot of things. I will update the pull request with the changes I mentioned earlier but without any changes to the tests for now (which might be required). I will implement them later.

You can still comment any of the changes I already did following your previous remarks.

@alexmv

alexmv commented on zerver/lib/send_email.py:
```diff
@@ -231,6 +246,23 @@ def send_email_from_dict(email_dict: Mapping[str, Any]) -> None:
+def initialize_connection(connection: Optional[BaseEmailBackend]) -> BaseEmailBackend:
+    if not connection:
+        connection = get_connection()
+    connection.open()
+    # Ensures that we don't return a connection that has been closed by the mail server
+    if isinstance(connection, EmailBackend):
+        try:
+            status = connection.connection.noop()[0]
+        except Exception:
+            status = -1
+        if status != 250:
+            # Try again; calling ourselves without passing down the 'connection' will make a new low-level SMTP connection
+            return initialize_connection()
+    return connection
```
Looking at my suggested code with fresh eyes, I think it has the failure mode that if a `noop` is _unsuccessful after a new connection is opened_ then we'll recurse infinitely here, busily making and trying to `noop` connections.

The other problem is that, from briefly poking at it, `@backoff.on_exception` re-calls its decorated function with the same parameters.  And a `connection` which has already failed its low-level noop, but still thinks it's open, is going to not re-open the TCP connection, and fail repeatedly.  Recursing is also going to open us up to the problem where the inner call retries 3 times, then re-raises -- whereupon the _outer_ call will retry 3 times.

Ditching the recursive calls, and being explicit about open/close, is I think enough here to fix this:
```suggestion
    if connection.open():
        # If it's a new connection, no need to no-op to check connectivity
        return connection
    # No-op to ensures that we don't return a connection that has been closed by the mail server
    if isinstance(connection, EmailBackend):
        try:
            status = connection.connection.noop()[0]
        except Exception:
            status = -1
        if status != 250:
            # Close and connect again.
            connection.close()
            connection.open()
    return connection
```

pletinckxc replied:
Indeed these changes seem to be required for this function to work properly. Modified.

alexmv commented on zerver/worker/queue_processors.py:
```diff
@@ -607,17 +609,33 @@ def maybe_send_batched_emails(self) -> None:
+    def stop(self) -> None:
+        try:
+            self.connection.close()
+            super().stop()
+        except smtplib.SMTPException:
+            super().stop()
```
```suggestion
        finally:
            super().stop()
```

alexmv commented on zerver/worker/queue_processors.py:
```diff
@@ -607,17 +609,33 @@ def maybe_send_batched_emails(self) -> None:
+        for copied_event in copied_events:
+            if "failed_tries" in copied_event:
+                del copied_event["failed_tries"]
+            copied_event["connection"] = self.connection
```
Before we pass the connection down, we should make sure that it's "fresh."
```suggestion
            self.connection = initialize_connection(self.connection)
            copied_event["connection"] = self.connection
```

alexmv reviewed (commented):
This is looking good!  Some inline comments where I led you astray before.

Now let's write some tests for it. :)

https://zulip.readthedocs.io/en/latest/testing/testing.html#internet-access-inside-test-suites and https://zulip.readthedocs.io/en/latest/testing/testing-with-django.html#testing-with-mocks may be helpful.

pletinckxc commented:
@alexmv I pushed some modifications to account for your second wave of comments and I started to work on the tests but I can't find a way to test this code snippet : 

```
if isinstance(connection, EmailBackend):
        try:
            status = connection.connection.noop()[0]
        except Exception:
            status = -1
        if status != 250:
            # Close and connect again.
            connection.close()
            connection.open()
    return connection
```
I tried many different possibilities with the mock module but I can't find one that is able to test this correctly. My first problem occurs at the first line : the connection should be of type `smtp.EmailBackend` to enter the if statement. The best way I found to do this in the tests is to create an object of this class and to mock its methods with `patch.object`. This is necessary because in the tests any calls to `get_connection` create a `locmem.EmailBackend` object and not an `smtp.EmailBackend` one. Is it an elegant way of doing it ?

My second issue is that I can't find a way to mock the `connection.connection.noop()` call. When I instantiate the `smtp.EmailBackend` class its connection is set to `None` and thus when I try 
`mock.patch.object(smtp.EmailBackend.connection, "noop", side_effect=Exception)` an `AttributeError` is thrown : 
`AttributeError: None does not have the attribute 'noop'`. This is logical but I can't call `smtp.EmailBackend.open()` as it uses an internet connection so is there a way to initialize the `connection` attribute in order to be able to mock `noop` ?

alexmv commented:
On the mocking front, I think `MagicMock(spec=...)` may help.  Here's a quick demo:

```
from unittest.mock import MagicMock
import smtplib
from django.core.mail.backends.smtp import EmailBackend

c = MagicMock(spec=EmailBackend)
c.connection = MagicMock(spec=smtplib.SMTP)

assert isinstance(c, EmailBackend)

c.open.return_value = False
c.connection.noop.return_value = [250]  # Or give it a side effect, to test the `except`
initialize_connection(c)

assert c.open.call_count == 1
assert c.connection.noop.call_count == 1
```

You can mock out `get_connection` to return that, for he case where you call `initialize_connection(None)`.

pletinckxc commented on zerver/lib/test_helpers.py:
```diff
@@ -694,8 +694,14 @@ def verify_serialize(
         event: Dict[str, object],
         processor: Optional[Callable[[object], None]] = None,
     ) -> None:
+        # orjson can't serialize EmailBackend objects;
+        connection = None
+        if "connection" in event:
+            connection = event.pop("connection")
         marshalled_event = orjson.loads(orjson.dumps(event))
         assert marshalled_event == event
+        if connection:
+            event["connection"] = connection
```
See my answer below.

pletinckxc commented on zerver/worker/queue_processors.py:
```diff
@@ -607,18 +612,35 @@ def maybe_send_batched_emails(self) -> None:
+        for event in events:
+            # Refresh connection if needed
+            self.connection = initialize_connection(self.connection)
+            event["connection"] = self.connection
```
This line is necessary in my opinion as the connection attribute is needed by the `send_email` function in `send_email.py` to actually send the email through it. However it is not at the right place. The issue that you mentioned above is linked to the `retry_send_email_failures` decorator that pushes the event back to the queue three times before giving up. This means that we can't insert the connection attribute in the event but rather have to copy it before passing it to `send_email` with the connection attribute. As the event was already copied to avoid passing `failed_tries` to `send_email_from_dict` I used the same copy to include the connection without impacting the serialization of the event that will be pushed into the queue.

pletinckxc commented:
Thank you once again for your feedback. I updated the changes and implemented the test for `initialize_connection`. I don't think we need any additional tests as `send_email` in `send_email.py` and the `EmailSendingWorker` class are tested through numerous other test suites that use them in the process.

I remain open to any additional comments if you consider that the test suite I implemented is not sufficient.

@alexmv

alexmv commented:
Merged, with minor edits, as 9afde790c6a1855de1714378b1ee3e467a7b0fb1!  Thanks for this fix, @pletinckxc!

timabbott commented:
This has at least one bug: https://chat.zulip.org/#narrow/stream/3-backend/topic/SMTP.20exceptions/near/1170218

I haven't done any further investigation.
